### PR TITLE
Dynamic index vesion control

### PR DESCRIPTION
**Pull request by foxspy**
issue: #27177 

Use cluster session to determine the index version dynamically

### CODE REVIEW

File: internal/core/src/index/Utils.cpp [hunk 1]
```diff
@@ -120,11 +120,12 @@ GetIndexTypeFromConfig(const Config& config) {
     return index_type.value();
 }
 
-std::string
+int32_t
```
Comment (longjiquan) [inline internal/core/src/index/Utils.cpp:123]: Please use the alias `IndexVersion` instead.

File: internal/core/src/index/Utils.h [hunk 2]
```diff
@@ -100,7 +100,7 @@ GetMetricTypeFromConfig(const Config& config);
 std::string
 GetIndexTypeFromConfig(const Config& config);
 
-std::string
+int32_t
```
Comment (longjiquan) [inline internal/core/src/index/Utils.h:103]: Ditto.

File: internal/core/src/index/VectorIndex.h [hunk 3]
```diff
@@ -91,11 +91,11 @@ class VectorIndex : public IndexBase {
     }
 
     void
-    CheckCompatible(const std::string& version) {
+    CheckCompatible(const int64_t& version) {
```
Comment (longjiquan) [inline internal/core/src/index/VectorIndex.h:94]: Ditto.

File: internal/core/thirdparty/knowhere/CMakeLists.txt [hunk 4]
```diff
@@ -11,8 +11,8 @@
 # or implied. See the License for the specific language governing permissions and limitations under the License.
 #-------------------------------------------------------------------------------
 
-set( KNOWHERE_VERSION b724230 )
-set( GIT_REPOSITORY  "https://github.com/zilliztech/knowhere.git")
+set( KNOWHERE_VERSION update_knowhere_version )
+set( GIT_REPOSITORY  "https://github.com/foxspy/knowhere.git")
```
Comment (longjiquan) [inline internal/core/thirdparty/knowhere/CMakeLists.txt:15]: Correct this when the knowhere is ready.

File: internal/datacoord/index_engine_version_manager.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+		return 0
+	}
+
+	current := int32(math.MaxInt32)
+	for _, version := range m.versions {
+		if version.CurrentIndexVersion < current {
+			current = version.CurrentIndexVersion
+		}
+	}
+	log.Info("getIndexEngineVersion", zap.Int32("minimalVersion", current))
```
Comment (longjiquan) [inline internal/datacoord/index_engine_version_manager.go:76]: Log message should be `current`.

File: internal/datacoord/index_service.go [hunk 6]
```diff
@@ -51,24 +51,25 @@ func (s *Server) startIndexService(ctx context.Context) {
 }
 
 func (s *Server) createIndexForSegment(segment *SegmentInfo, indexID UniqueID) error {
-	indexEngineVersion := Params.CommonCfg.IndexEngineVersion.GetValue()
+	indexEngineVersion := s.IndexEngineVersionManager.GetCurrentIndexEngineVersion()
 	log.Info("create index for segment", zap.Int64("segmentID", segment.ID), zap.Int64("indexID", indexID),
-		zap.String("index_engine_version", indexEngineVersion),
+		zap.Int32("index_engine_version", indexEngineVersion),
```
Comment (longjiquan) [inline internal/datacoord/index_service.go:56]: Please also log the minimal version.

File: internal/util/sessionutil/session_util.go [hunk 7]
```diff
@@ -143,14 +157,15 @@ func (s *Session) apply(opts ...SessionOption) {
...
-		Version     string            `json:"Version"`
-		LeaseID     *clientv3.LeaseID `json:"LeaseID,omitempty"`
+		ServerID           int64  `json:"ServerID,omitempty"`
+		ServerName         string `json:"ServerName,omitempty"`
+		Address            string `json:"Address,omitempty"`
+		Exclusive          bool   `json:"Exclusive,omitempty"`
+		Stopping           bool   `json:"Stopping,omitempty"`
+		TriggerKill        bool
+		Version            string            `json:"Version"`
+		IndexEngineVersion string            `json:"IndexEngineVersion,omitempty"`
```
Comment (longjiquan) [inline internal/util/sessionutil/session_util.go:167]: Struct type can be also marshaled.

**Review by longjiquan — CHANGES_REQUESTED** [review]

File: internal/datacoord/index_service.go (hunk 6 above)
Comment (longjiquan) [reply internal/datacoord/index_service.go:56]: Meanwhile, do not call the function twice.

Comment (foxspy) [reply internal/datacoord/index_service.go:56]: done

File: internal/core/thirdparty/knowhere/CMakeLists.txt (hunk 4 above)
Comment (foxspy) [reply internal/core/thirdparty/knowhere/CMakeLists.txt]: done

File: internal/core/src/index/Utils.cpp (hunk 1 above)
Comment (foxspy) [reply internal/core/src/index/Utils.cpp]: done

**Pull request merged**